src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs

```diff
@@ -1088,83 +1090,114 @@ internal async Task InvokeResponse(InvocationResponse invokeResponse)
                     {
-                        context.ResultSource.TrySetException(responseEx);
+                        ScriptInvocationResult result = await CreateScriptInvocationResult(invokeResponse);
+                        context.ResultSource.SetResult(result);
+
+                        _metricsLogger.LogEvent(_workerInvocationSucccededMetric);
                     }
-                    finally
+                    else
                     {
-                        // Free memory allocated by the host (for input bindings) which was needed only for the duration of this invocation
-                        if (!_sharedMemoryManager.TryFreeSharedMemoryMapsForInvocation(invokeResponse.InvocationId))
-                        {
-                            _workerChannelLogger.LogWarning("Cannot free all shared memory resources for invocation: {invocationId}", invokeResponse.InvocationId);
-                        }
+                        var rpcException = invokeResponse.Result.GetRpcException(userCodeExceptionHandlingEnabled);
+                        context.ResultSource.SetException(rpcException);
 
-                        // List of shared memory maps that were produced by the worker (for output bindings)
-                        IList<string> outputMaps = GetOutputMaps(invokeResponse.OutputData);
-                        if (outputMaps.Count > 0)
-                        {
-                            // If this invocation was using any shared memory maps produced by the worker, close them to free memory
-                            SendCloseSharedMemoryResourcesForInvocationRequest(outputMaps);
-                        }
+                        _metricsLogger.LogEvent(_workerInvocationFailedMetric);
                     }
+                }
+                catch (Exception exc)
+                {
+                    context.ResultSource.TrySetException(exc);
+                }
+                finally
+                {
+                    TryCloseSharedMemoryResources(invokeResponse);
 
                     invocation.Dispose();
                 }
-                else
+            }
+        }
+
+        private async Task<ScriptInvocationResult> CreateScriptInvocationResult(InvocationResponse invokeResponse)
+        {
+            IDictionary<string, object> bindingsDictionary = await invokeResponse.OutputData
+                .ToDictionaryAsync(binding => binding.Name, binding => GetBindingDataAsync(binding, invokeResponse.InvocationId));
+
+            var result = new ScriptInvocationResult()
+            {
+                Outputs = bindingsDictionary,
+                Return = invokeResponse?.ReturnValue?.ToObject()
+            };
+
+            return result;
+        }
+
+        private void TryCloseSharedMemoryResources(InvocationResponse invokeResponse)
+        {
+            try
+            {
+                if (!IsSharedMemoryDataTransferEnabled())
```

Comment:
Performing checks now to make sure this doesn't run when not needed.

Reply:
Do we need to execute this method for every invocation response? Looks like it does some try parse + logging inside it. Could we use `_isSharedMemoryDataTransferEnabled` variable?

Reply:
Addressed this. Checking for the variable now, but also updated the method to make that API a bit clearer and avoid having it used the same way again. Also removed state dependency there.